Add initial models

ZachCMP opened the pull request:
https://www.pivotaltracker.com/story/show/184496458

ZachCMP committed:
Add initial models

ZachCMP committed:
Merge branch 'release-67' into 184496458-access-control-models

ZachCMP commented:
@eanders @gigxz I tried my best to rework things so that the tests pass, but the logic probably needs more work and consideration. The complexity that users now have any number of roles across any number of groups breaks some of the role conventions we had before. Also operations like `role.add(user)` aren't really possible any more without knowing about what group we want to add them to, so I've mostly removed those and moved them to the `Hmis::AccessControl` model where we can actually infer those things.

I've also put some code into the tests to handle some of these changes for the moment, but we should go back and revise the tests once we're clear on how we want to assign roles/groups/ACLs overall.

gigxz commented on db/migrate/20230217151359_add_access_control_list_models.rb:
```diff
@@ -0,0 +1,17 @@
+class AddAccessControlListModels < ActiveRecord::Migration[6.1]
+  def change
+    create_table :hmis_access_controls do |t|
```
There should be a `name` on the access control list

eanders replied:
I don't think it' needs a name.  The name should be the name of the role and Group that are associated with it.  That should prevent duplicate access control lists from being created.

gigxz replied:
Oh ok, that makes sense

gigxz commented on drivers/hmis/app/models/hmis/access_control.rb:
```diff
@@ -0,0 +1,25 @@
+
+  def add(users)
+    (self.users + Array.wrap(users)).uniq.each do |user|
+      user_access_controls.create!(user_id: user.id)
```
Won't this create duplicate entries, since we're calling `create!` for users that have already been added to the list?

I think that could be fixed by using `find_or_create_by`. I don't see the need to iterate over `self.users` here because those already exist.

gigxz commented on drivers/hmis/app/models/hmis/access_control.rb:
```diff
@@ -0,0 +1,25 @@
+    end
+  end
+
+  def remove(users)
+    user_access_controls.where(user_id: users.pluck(:id)).destroy_all
+  end
```
Can we update this to use `Array.wrap(users)` so that it's possible to pass a single user to `remove`

eanders replied:
You can't call `pluck` on an array.  Users will need to be a relation.  Alternatively, you could accept an array of User IDs (or a single ID) and do the pluck in the calling method.

gigxz replied:
You can https://apidock.com/rails/Enumerable/pluck but that is probably less performant to turn the relation into an array before plucking

ZachCMP replied:
@gigxz So would you like me to add the `Array.wrap` part, or just leave it for now?

gigxz replied:
I don't feel strongly, you can leave it for now

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -54,6 +57,23 @@ def load_effective_permissions
       send(permission)
     end
 
+    define_method("#{permission}_for?") do |entity|
+      joins_association = nil
+      joins_association = :projects if entity.is_a?(Hmis::Hud::Project)
+      joins_association = :organizations if entity.is_a?(Hmis::Hud::Organization)
+
+      raise "Invalid entity '#{entity.class.name}'" unless joins_association.present?
+      return false unless send("#{permission}?")
+
+      access_groups.
+        merge(
+          Hmis::AccessGroup.joins(:access_controls).
+            where(id: Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id).pluck(:access_group_id)).
+            merge(Hmis::AccessControl.where(role_id: roles.where(permission => true).pluck(:id))),
+        )
+        .exists?
+    end
```
I _think_ this logic makes sense to me, but I'm finding it confusing to read.

Would the below logic achieve the same result?

```
     # All access groups that include the specified entity
     access_group_ids = Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id).pluck(:access_group_id)
    
     # All roles that include the specified permission
     role_ids = Hmis::Role.where(permission => true)

     access_controls.where(access_group_id: access_group_ids, role_id: role_ids).exists?
```

gigxz replied:
Actually, there is an issue here. `Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id)` isn't the right way to find _any GroupViewableEntity that grants access to this <project|organization>_. It will only return GroupViewableEntities that are explicitly assigned to that entity. It incorrectly excludes group viewable entities that grant access to a project by way of an organization, data source, or project group.


In other words: you should be able to call `user.permission_for?(organization)` and `user.permission_for?(project)`  and have it return `true` if the user is part of an ACL that belongs to a _group_ that grants access to the _data source_ that the specified project/organization belong to.

Maybe a clean way to do this would be to add scopes for `:includes_project` and `:includes_organization` so we can say _"give me all the GroupViewableEntities that grant access to this record."_ Then we could do something like:

```
# All access groups that include the specified project
access_group_ids = Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Project)
access_group_ids = Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Organization)
```

gigxz replied:
We may want to chat with @awisema to get some feedback on this. I don't know if we do anything similar with GroupViewableEntities in the warehouse. The closest thing I can think of is the Project `viewable_by` scope which evaluates all entity levels to find all viewable projects.

ZachCMP committed:
Merge branch 'release-67' into 184496458-access-control-models

ZachCMP committed:
Add test and fix authorization logic

ZachCMP commented:
@gigxz I updated the logic to work with projects, organizations, data sources, and project groups, and I added an access control test to verify the logic is working as expected. Take a look and let me know what else you need from me on this.

gigxz commented on drivers/hmis/app/models/hmis/group_viewable_entity.rb:
```diff
@@ -10,5 +10,27 @@ class GroupViewableEntity < GrdaWarehouse::GroupViewableEntity
+      where(entity_type: project.class.name, entity_id: project.id).
+        or(includes_data_source(project.data_source)).
+        or(includes_organization(project.organization)).
+        or(includes_data_source(project.organization.data_source)).
```
Line 19 can be removed - the project data source and organization data source are the same

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -65,10 +65,15 @@ def load_effective_permissions
 
+      access_group_ids = Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Project)
+      access_group_ids = Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Organization)
+      access_group_ids = Hmis::GroupViewableEntity.includes_data_source(entity).pluck(:access_group_id) if entity.is_a?(GrdaWarehouse::DataSource)
+      access_group_ids = Hmis::GroupViewableEntity.includes_project_access_group(entity).pluck(:access_group_id) if entity.is_a?(GrdaWarehouse::ProjectAccessGroup)
```
These 2 can be removed. We will only ever need to check access at the project and org level. 90% of the time will be using project. Organization is needed for the "can edit organization" type permissions. There is no Data Source or Project Group editing in the HMIS, so we shouldn't need those.

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -65,10 +65,15 @@ def load_effective_permissions
+            where(id: access_group_ids).
             merge(Hmis::AccessControl.where(role_id: roles.where(permission => true).pluck(:id))),
         )
         .exists?
```
@ZachCMP for my own understanding here, can you explain why this nested merging approach is preferable to just using a `where` clause? I think they have the same result, but I could be wrong.
```
access_group_ids = ...
role_ids = roles.where(permission => true).pluck(:id)

access_controls.where(access_group_id: access_group_ids, role_id: role_ids).exists?
```

ZachCMP replied:
@gigxz I changed the code to what you're suggesting here and it seems to work fine. I worked up the initial code with @eanders and followed his lead on how to assemble the query logic, so I probably couldn't explain any better than him why the merging would be needed in this case.

Since the code you suggest is cleaner and also makes the tests pass, I've gone ahead and switched to that.

gigxz commented on drivers/hmis/spec/models/hmis/access_control.rb:
```diff
@@ -0,0 +1,91 @@
+require_relative '../../requests/hmis/login_and_permissions'
+require_relative '../../support/hmis_base_setup'
+
+RSpec.describe Hmis::AccessControl, type: :model do
```
Tests look good 👍

gigxz reviewed (commented):
Thanks @ZachCMP! This looks great. There are a few small things to address, see comments. There is also an unaddressed comment from my previous review about duplication.

eanders commented on drivers/hmis/app/models/hmis/access_control.rb:
```diff
@@ -0,0 +1,25 @@
+  has_many :users, through: :user_access_controls
+
+  def add(users)
+    (self.users + Array.wrap(users)).uniq.each do |user|
+      user_access_controls.create!(user_id: user.id)
+    end
```
```suggestion
    Array.wrap(users).uniq.each do |user|
      user_access_controls.where(user_id: user.id).first_or_create!
    end
```

eanders commented on drivers/hmis/app/models/hmis/access_group.rb:
```diff
@@ -12,39 +12,40 @@ class Hmis::AccessGroup < ApplicationRecord
-  # If you can edit, you can also view
   scope :viewable, -> do
-    where(scope: ['view', 'edit'])
+    joins(:roles).merge(Hmis::Role.with_viewable_permissions)
   end
 
   scope :editable, -> do
-    where(scope: ['edit'])
+    joins(:roles).merge(Hmis::Role.with_editable_permissions)
   end
```
We may want the distinction between viewable and editable at this level, but it seems unnecessary to me.  Is this so we can more easily determine if the page should be read-only or is this just left over from the previous mechanism?

gigxz replied:
I think it's just left over, I don't see why we would need it. Each mutation should eventually be checking for specific permission(s).

ZachCMP replied:
Agreed that this is not a necessary distinction and is better abandoned. IIRC I've left it in here to preserve the functionality of the `editable` and `viewable` methods on `Hmis::User`, and perhaps in other places. I didn't feel like I had a deep enough understanding of the intent of those methods to make decisions on their functionality without input from you all, so I erred on the side of preserving functionality that was already there.

Overall I think dropping the editable/viewable distinction in all cases in favor of checking for the presence of specific permissions is best, but I'm open to being wrong about that. As with the above, I wanted to get input from you all before making that decision.

ZachCMP replied:
Based on [conversation in slack](https://greenriver.slack.com/archives/C033YNSN0Q1/p1677531465874729) on 2/27/23, we're going to leave the `editable` and `viewable` distinction until we migrate completely over to ACLs

eanders commented on drivers/hmis/app/models/hmis/role.rb:
```diff
@@ -8,8 +8,37 @@ class Hmis::Role < ::ApplicationRecord
+    scope
+  end
+
+  scope :with_editable_permissions, -> do
+    with_any_permissions(
+      :can_delete_assigned_project_data,
+      :can_administer_hmis,
+      :can_delete_enrollments,
+    )
+  end
+
+  scope :with_viewable_permissions, -> do
+    with_any_permissions(
+      :can_view_full_ssn,
+      :can_view_clients,
+    )
+  end
```
I think it'll be better long-term to decorate the `permissions_with_descriptions` below with if the permission is editable or viewable, something like:
```
can_view_full_ssn: {
        description: 'Allow the user to see client\'s full SSN.',
        administrative: false,
        access: [:viewable],
        categories: [
          'Client Details',
        ],
      },
```
and then make a method that returns an array of the keys for editable or viewable something like:
```
def permissions_for_access(access_level)
  permissions_with_descriptions.select { |m| m[: access].include?(access_level) }.keys
end
```
That will centralize the identification of access and allow a permission to be both a view and edit (or other type of access in the future.)

eanders commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -54,6 +57,28 @@ def load_effective_permissions
+      raise "Invalid entity '#{entity.class.name}'" unless joins_association.present?
+      return false unless send("#{permission}?")
+
+      access_group_ids = Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Project)
+      access_group_ids = Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Organization)
```
As @gigxz we only need project and organization.  Also, can we adjust the syntax to be either an if/else or a case statement?  I'm not finding the many sequential trailing ifs very easy to follow.

ZachCMP committed:
Update drivers/hmis/app/models/hmis/access_control.rb

Co-authored-by: Elliot <elliot@greenriver.org>

ZachCMP committed:
Merge branch 'release-67' into 184496458-access-control-models

ZachCMP committed:
Fixup PR feedback

ZachCMP commented:
@gigxz @eanders All your feedback should be addressed now. I've also taken it out of draft mode.

ZachCMP committed:
Add graphql changes

ZachCMP committed:
Checkout base warehouse structure

ZachCMP committed:
Update schema

ZachCMP committed:
Update schema

ZachCMP committed:
cleanup user editable scopes

eanders commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -54,6 +57,24 @@ def load_effective_permissions
+
+      access_group_ids = nil
+
+      if entity.is_a?(Hmis::Hud::Project)
+        access_group_ids = Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id)
+      elsif entity.is_a?(Hmis::Hud::Organization)
+        access_group_ids = Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id)
+      end
```
```suggestion
       access_group_ids = if entity.is_a?(Hmis::Hud::Project)
       Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id)
      elsif entity.is_a?(Hmis::Hud::Organization)
        Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id)
      end
```

gigxz commented:
This looks good to me once tests are fixed @ZachCMP

ZachCMP committed:
Update drivers/hmis/app/models/hmis/user.rb

Co-authored-by: Elliot <elliot@greenriver.org>

ZachCMP committed:
Fix tests

ZachCMP commented:
@gigxz The tests should be fixed in #2840, and should pass once that branch is merged. I'm waiting to see if the tests pass in that PR right now.

ZachCMP committed:
Merge remote-tracking branch 'origin/184496458-access-control-models' into 184497332-graphql-permissions

ZachCMP committed:
Fix scope

ZachCMP committed:
Merge pull request #2840 from greenriver/184497332-graphql-permissions

Add graphql changes

ZachCMP committed:
Merge branch 'release-67' into 184496458-access-control-models

ZachCMP committed:
Add new permissions

ZachCMP commented:
@gigxz Tests are now passing. Also I've done the backend work for the [9 MVP permissions](https://www.pivotaltracker.com/story/show/184573644) and included them here, which are done exactly like the other permissions. Please let me know if you have any questions or concerns about that

ZachCMP commented:
Also added work for resolving root-level permissions on query: https://www.pivotaltracker.com/story/show/184497338

gigxz commented on drivers/hmis/app/models/hmis/group_viewable_entity.rb:
```diff
@@ -0,0 +1,54 @@
+    scope :data_sources, -> { where(entity_type: GrdaWarehouse::DataSource.sti_name) }
+    scope :project_access_groups, -> { where(entity_type: GrdaWarehouse::ProjectAccessGroup.sti_name) }
+
+    scope :includes_project, ->(project) do
+      where(entity_type: project.class.name, entity_id: project.id).
+        or(includes_data_source(project.data_source)).
+        or(includes_organization(project.organization)).
+        or(where(entity_type: GrdaWarehouse::ProjectAccessGroup.name, entity_id: project.project_groups.pluck(:id)))
+    end
```
@eanders

eanders replied:
Maybe this should be adjusted to match the `includes_data_source` pattern, so `included_in_project_groups(project.project_groups)` or similar?

ZachCMP replied:
fixed

gigxz commented on drivers/hmis/app/models/hmis/group_viewable_entity.rb:
```diff
@@ -0,0 +1,54 @@
+      where(entity_type: data_source.class.name, entity_id: data_source.id)
+    end
+
+    scope :includes_entity, ->(entity) do
```
Is this used anywhere?

ZachCMP replied:
No I think we're only using `includes_project` directly at the moment. But I see a use case for having a method that accepts any kind of entity that can have viewable entities, so think it's fine to leave in if we think we might want it at some point. Though I'm happy to remove it if you'd prefer. Let me know if you'd like me to do that

eanders replied:
Maybe we should upgrade the use of `includes_project` to `includes_entity` since it should work (I think) and would make future uses more obvious.

ZachCMP replied:
@eanders Do you want me to do that now?

ZachCMP replied:
@eanders I just went ahead and made that change, since it allowed me to remove a chunk of unnecessary code from the user model, which is a win in my book

gigxz commented on drivers/hmis/app/models/hmis/user.rb:
```diff
@@ -7,18 +7,21 @@
 # u = Hmis::User.first; u.hmis_data_source_id = 3
-# u.user_hmis_data_sources_roles.create(role: r, data_source_id: u.hmis_data_source_id)
+# g = Hmis::AccessGroup.create(name: 'test')
+# ac = u.access_controls.create(role: r, group: g)
```
```suggestion
# ac = u.access_controls.create(role: r, access_group: g)
```

ZachCMP committed:
Update drivers/hmis/app/models/hmis/user.rb

Co-authored-by: Gig <gig@greenriver.org>

gigxz commented:
CI is failing on the schema check @ZachCMP

ZachCMP committed:
PR feedback

ZachCMP committed:
Update schema

gigxz commented on drivers/hmis/app/models/hmis/access_group.rb:
```diff
@@ -135,7 +124,7 @@ def associated_by(associations:)
         end
       when :project_access_group
-        project_access_groups.preload(:projects).map do |pag|
+        project_access_access_.preload(:projects).map do |pag|
```
I'm getting errors here, this isn't defined

ZachCMP replied:
See below. If you're adding back the `project_access_groups` scope in your branch, probably best to just fix this in your branch too. it should be `project_access_groups`

gigxz replied:
its fixed on my branch

gigxz commented on drivers/hmis/app/models/hmis/access_group.rb:
```diff
@@ -12,39 +12,40 @@ class Hmis::AccessGroup < ApplicationRecord
-  has_many :data_sources, through: :group_viewable_entities, source: :entity, source_type: 'GrdaWarehouse::DataSource'
-  has_many :organizations, through: :group_viewable_entities, source: :entity, source_type: 'GrdaWarehouse::Hud::Organization'
-  has_many :projects, through: :group_viewable_entities, source: :entity, source_type: 'GrdaWarehouse::Hud::Project'
-  has_many :project_access_groups, through: :group_viewable_entities, source: :entity, source_type: 'GrdaWarehouse::ProjectAccessGroup'
```
`has_many :project_access_groups` and `has_many :data_sources` should not have been removed. I'm adding them back in my PR for the warehouse frontend part, so no action needed here

ZachCMP replied:
Okay, sounds good. This is related to the issue above, where the undefined method should be `project_access_groups`

eanders commented on drivers/hmis/app/models/hmis/role.rb:
```diff
@@ -44,54 +66,124 @@ def self.ensure_permissions_exist
   end
 
+  def self.permissions_for_access(access)
+    permissions_with_descriptions.select { |_k, attrs| access == :viewable ? attrs[:access] == [:viewable] : attrs[:access].include?(access) }.keys
```
why not?
```suggestion
    permissions_with_descriptions.select { |_k, attrs| attrs[:access].include?(access) }.keys
```

ZachCMP replied:
You are correct, that's the right way to do it. However, this will make tests fail if a permission grants both `viewable` and `editable` access, as `can_administer_hmis` does. I'm going to adjust it so that `can_administer_hmis` only has `editable` access, which should fix the issue with the tests as that will make all accesses mutually exclusive. We can change this later once we're no longer relying on the `editable_by` and `viewable_by` scopes.

ZachCMP replied:
Fixed

ZachCMP committed:
Add ACL migration; PR feedback

ZachCMP commented:
@gigxz @eanders I added work for the [ACL migration issue](https://www.pivotaltracker.com/story/show/184575016), and addressed a bunch more of the PR feedback

gigxz commented on db/migrate/20230217151360_migrate_access_groups_to_access_control_lists.rb:
```diff
@@ -0,0 +1,24 @@
+        # Users that have this role and are in this group
+        agm_scope = Hmis::AccessGroupMember.where(access_group_id: access_group.id).pluck(:user_id)
+        ushdr_scope = Hmis::UserHmisDataSourceRole.where(role_id: role.id)
+        users = Hmis::User.where(id: agm_scope.pluck(:user_id)).merge(Hmis::User.where(user_id: ushdr_scope.pluck(:user_id)))
```
Does this work? `pluck(:user_id)` is being called twice on `agm_scope`

ZachCMP replied:
Good catch! Fixed

gigxz committed:
Admin interface for HMIS ACLs

gigxz committed:
Merge branch '184496458-access-control-models' into gig/acl-warehouse-interface

gigxz committed:
Fix migration

ZachCMP committed:
Fixup migration & user scope

gigxz committed:
Finish updates to admin ui

gigxz committed:
Add back role editing

gigxz committed:
Merge branch '184496458-access-control-models' into gig/acl-warehouse-interface

gigxz committed:
Bad merge

gigxz committed:
Rm link to edit

eanders committed:
Migration cleanup

eanders committed:
Merge pull request #2845 from greenriver/gig/acl-warehouse-interface

Admin interface for managing Access Control Lists

eanders commented:
I'm good with this when @gigxz is.

gigxz reviewed (approved):
merge away!
